src/hotspot/share/gc/g1/g1CardSet.cpp

```diff
@@ -44,36 +44,54 @@
 
 G1CardSet::CardSetPtr G1CardSet::FullCardSet = (G1CardSet::CardSetPtr)-1;
 
+static uint default_log2_card_region_per_region() {
+  uint log2_card_region_per_heap_region = 0;
+
+  const uint card_container_limit = G1CardSetContainer::LogCardsPerRegionLimit;
+  if (card_container_limit < (uint)HeapRegion::LogCardsPerRegion) {
+    log2_card_region_per_heap_region = (uint)HeapRegion::LogCardsPerRegion - card_container_limit;
+  }
+
+  return log2_card_region_per_heap_region;
+}
+
 G1CardSetConfiguration::G1CardSetConfiguration() :
   G1CardSetConfiguration(HeapRegion::LogCardsPerRegion,                             /* inline_ptr_bits_per_card */
                          G1RemSetArrayOfCardsEntries,                               /* num_cards_in_array */
                          (double)G1RemSetCoarsenHowlBitmapToHowlFullPercent / 100,  /* cards_in_bitmap_threshold_percent */
                          G1RemSetHowlNumBuckets,                                    /* num_buckets_in_howl */
                          (double)G1RemSetCoarsenHowlToFullPercent / 100,            /* cards_in_howl_threshold_percent */
-                         (uint)HeapRegion::CardsPerRegion)                          /* max_cards_in_cardset */
-                         { }
+                         (uint)HeapRegion::CardsPerRegion,                          /* max_cards_in_cardset */
+                         default_log2_card_region_per_region())                     /* log2_card_region_per_region */
+{
+  assert((_log2_card_region_per_heap_region + _log2_card_region_size) == (uint)HeapRegion::LogCardsPerRegion,
```

Comment:
I suggest `_log2_card_region_size -> _log2_cards_per_card_region`; then the assertion becomes more consistent, IMO.
`assert((_log2_card_region_per_heap_region + _log2_cards_per_card_region) == (uint)HeapRegion::LogCardsPerRegion)`, which, in math form, is
`(card_region / heap_region) x (cards / card_region) == cards / (heap_)region`

Reply:
Implemented. I'll push in around 7-8 hours unless somebody objects.